src/utils/rlp.cairo

```diff
@@ -1,69 +1,137 @@
+        let is_le_183 = is_le(prefix, 183);  // a max 55 bytes long string
         if (is_le_183 != FALSE) {
-            let string_len = byte - 128;
-            assert [items] = Item(data_len=string_len, data=data, is_list=0);
-            return decode(
-                data_len=data_len - 1 - string_len, data=data + string_len, items=items + Item.SIZE
-            );
+            return (0, 1, prefix - 128);
         }
 
-        let is_le_191 = is_le(byte, 191);  // string longer than 55 bytes
+        let is_le_191 = is_le(prefix, 191);  // string longer than 55 bytes
         if (is_le_191 != FALSE) {
-            local len_len = byte - 183;
-            let dlen = Helpers.bytes_to_felt(len_len, data);
-            let data = data + len_len;
-            assert [items] = Item(data_len=dlen, data=data, is_list=0);
-            return decode(
-                data_len=data_len - 1 - len_len - dlen, data=data + dlen, items=items + Item.SIZE
-            );
+            local len_bytes_count = prefix - 183;
+            let string_len = Helpers.bytes_to_felt(len_bytes_count, data + 1);
+            return (0, 1 + len_bytes_count, string_len);
         }
 
-        let is_le_247 = is_le(byte, 247);  // list 0-55 bytes long
+        let is_le_247 = is_le(prefix, 247);  // list 0-55 bytes long
         if (is_le_247 != FALSE) {
-            local list_len = byte - 192;
-            assert [items] = Item(data_len=list_len, data=data, is_list=1);
-            return decode(
-                data_len=data_len - 1 - list_len, data=data + list_len, items=items + Item.SIZE
+            local list_len = prefix - 192;
+            return (1, 1, list_len);
+        }
+
+        local len_bytes_count = prefix - 247;
+        let list_len = Helpers.bytes_to_felt(len_bytes_count, data + 1);
+        return (1, 1 + len_bytes_count, list_len);
+    }
+
+    // @notice Decodes a Recursive Length Prefix (RLP) encoded data.
+    // @notice This function decodes the RLP encoded data into a list of items.
+    // Each item is a struct containing the length of the data, the data itself, and a flag indicating whether the data is a list.
+    // The function first determines the type of the RLP data (string or list) and then processes it accordingly.
+    // If the data is a string, it is simply added to the items.
+    // If the data is a list, it is recursively decoded.
+    // After processing the first item, the function checks if there is more data to decode and if so,
+    // it recursively decodes the remaining data and adds the decoded items to the list of items.
+    // @param data_len The length of the data to decode.
+    // @param data The RLP encoded data.
+    // @return items_len The number of items decoded.
+    // @return items The decoded RLP items.
+    func decode{range_check_ptr}(data_len: felt, data: felt*) -> (items_len: felt, items: Item*) {
+        alloc_locals;
+        tempvar items_len: felt = 0;
+        let (items: Item*) = alloc();
```

Comment:
take this as input (arg or implicit arg) so that you don't need to `memcpy` at the end